src/resolve/resolved-hook.c

```diff
@@ -0,0 +1,825 @@
+        if (r < 0)
+                return r;
+
+        if (rcode < 0 || rcode >= _DNS_RCODE_MAX)
+                return json_log(variant, flags, SYNTHETIC_ERRNO(ERANGE), "JSON field '%s' contains an invalid DNS rcode.", strna(name));
+
+        *u = rcode;
+        return 0;
+}
+
+static int dispatch_answer(const char *name, sd_json_variant *variant, sd_json_dispatch_flags_t flags, void *userdata) {
+        DnsAnswer **a = ASSERT_PTR(userdata);
+        int r;
+
+        assert(variant);
+
+        if (sd_json_variant_is_null(variant)) {
+                *a = dns_answer_unref(*a);
+                return 0;
+        }
+
+        if (!sd_json_variant_is_array(variant))
+                return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "JSON field '%s' is not an array.", strna(name));
+
+        _cleanup_(dns_answer_unrefp) DnsAnswer *l = NULL;
+        sd_json_variant *e;
+        JSON_VARIANT_ARRAY_FOREACH(e, variant) {
+                if (!sd_json_variant_is_object(e))
+                        return json_log(e, flags, SYNTHETIC_ERRNO(EINVAL), "JSON array element is not an object");
+
+                _cleanup_(iovec_done) struct iovec iovec = {};
+                static const sd_json_dispatch_field dispatch_table[] = {
+                        { "raw", SD_JSON_VARIANT_STRING, json_dispatch_unbase64_iovec, 0, SD_JSON_MANDATORY },
+                        { "rr",  SD_JSON_VARIANT_OBJECT, NULL,                         0, 0                 },
+                        {}
+                };
+
+                r = sd_json_dispatch(e, dispatch_table, flags|SD_JSON_ALLOW_EXTENSIONS, &iovec);
+                if (r < 0)
+                        return r;
+
+                _cleanup_(dns_resource_record_unrefp) DnsResourceRecord *rr = NULL;
+                r = dns_resource_record_new_from_raw(&rr, iovec.iov_base, iovec.iov_len);
+                if (r < 0)
+                        return json_log(variant, flags, SYNTHETIC_ERRNO(EINVAL), "JSON field '%s' contains an invalid resource record.", strna(name));
+
+                if (dns_answer_add_extend(&l, rr, /* ifindex= */ 0, /* flags= */ 0, /* rrsig= */ NULL) < 0)
+                        return json_log_oom(e, flags);
+        }
+
+        dns_answer_unref(*a);
+        *a = TAKE_PTR(l);
+
+        return 0;
+}
+
+typedef struct QueryReplyParameters {
+        int rcode;
+        DnsAnswer *answer;
+} QueryReplyParameters;
+
+static void query_reply_parameters_done(QueryReplyParameters *p) {
+        assert(p);
+
+        p->answer = dns_answer_unref(p->answer);
+}
+
+static int on_query_reply(
+                sd_varlink *link,
+                sd_json_variant *parameters,
+                const char *error_id,
+                sd_varlink_reply_flags_t flags,
+                void *userdata) {
+
+        HookQueryCandidate *qc = ASSERT_PTR(userdata);
+        HookQuery *q = qc->query; /* save early in case we destroy 'qc' half-way through this function */
```

Comment:
You save this here but don't use the saved variable throughout the function. If you save this in a variable here you should use that variable everywhere throughout the function.